File: AOC_2022/Day 15/15.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <regex>

using namespace std;

class Solution
{
    public:
    map<pair<int, int>, pair<int, int>> sensorstobeacons;
    void init(ifstream& input){
        string line;
        // "Sensor at x=\d+, y=\d+: closest beacon is at x=\d+, y=\d+"
        regex pattern(R"(-?\d+)");
        while (getline(input, line))
        {
            vector<string> matches;
            sregex_iterator it(line.begin(), line.end(), pattern), end;
            while(it != end){
                matches.push_back(it->str());
                it++;
            }
            sensorstobeacons[{stoi(matches[0]), stoi(matches[1])}] = {stoi(matches[2]), stoi(matches[3])};
        }
    }

    // standard merge intervals Algo
    void mergeIntervals(vector<pair<int, int>>& intervals){
        sort(intervals.begin(), intervals.end());
        int index = 0;
        for(int i = 1; i < intervals.size(); i++){
            // if overlaps
            if(intervals[index].second >= intervals[i].first){
                intervals[index].second = max(intervals[index].second, intervals[i].second);
            }
            // if not overlaps
            else{
                index++;
                intervals[index] = intervals[i];
            }
        }
        intervals.resize(index+1);
    }

    void part1(){
        int y = 2000000, result = 0;
        vector<pair<int, int>> intervals;
        // for each sensor, find the manhattan distance, and then find the x distance
        for(auto it : sensorstobeacons){
            int manhattan = abs(it.first.first - it.second.first) + abs(it.first.second - it.second.second), y_dist = abs(it.first.second - y);
            // if y distance is greater than manhattan distance, then it is not the shortest distance to sensor
            if(y_dist > manhattan) continue;
            int x_dist = manhattan - y_dist;
            // add the x interval for the possible beacons  on y = 2000000 for each sensor
            intervals.push_back({it.first.first - x_dist, it.first.first + x_dist});
        }
        // merge the intervals
        mergeIntervals(intervals);
        pair<int,int> beacon_y;
        // find the beacon on y = 2000000
        for(auto it : sensorstobeacons){
            if(it.second.second == y){
                beacon_y = it.second;
                break;
            }
        }
        for(auto it : intervals){
            // add the length of the interval to the result
            result += it.second - it.first + 1;
            // if beacon is in the interval, then subtract 1 from the result
            if(beacon_y.second == y && beacon_y.first >= it.first && beacon_y.first <= it.second) result--;
        }
        cout<<"Part 1: "<<result<<"\n";
    }

    void part2(){
        long long result = 0;
        int mx = 4000000;
        // for each y, find the intervals of possible beacons
        for(int i = 0; i <= mx;i++){
            vector<pair<int, int>> intervals;
            // same as part 1
            for(auto it : sensorstobeacons){
                int manhattan = abs(it.first.first - it.second.first) + abs(it.first.second - it.second.second), y_dist = abs(it.first.second - i);
                if(y_dist > manhattan) continue;
                int x_dist = manhattan - y_dist;
                intervals.push_back({max(0,it.first.first - x_dist), min(mx, it.first.first + x_dist)});
            }
            mergeIntervals(intervals);
            // if there is only one interval, then it has a continuous range of beacons
            // else there is a discontinuity which is the the distress Beacon
            // Interval is of the form [x1, x2], [x2+2, x3]
            if (intervals.size() > 1){
                result = (intervals[0].second + 1);
                result *= mx;
                result += i;
                break;
            }
        }
        cout<<"Part 2: "<<result<<"\n";
    }

};

int main(){
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}

File: AOC_2022/Day 15/15.py
import re

class Solution:
    def __init__(self):
        self.sensorstobeacons = {}

    def init(self, lines):
        pattern = re.compile(r'-?\d+')
        for line in lines:
            matches = re.findall(pattern, line)
            sensor_coords = (int(matches[0]), int(matches[1]))
            beacon_coords = (int(matches[2]), int(matches[3]))
            self.sensorstobeacons[sensor_coords] = beacon_coords

    def merge_intervals(self, intervals):
        intervals.sort(key=lambda x: x[0])
        index = 0
        for i in range(1, len(intervals)):
            if intervals[index][1] >= intervals[i][0]:
                intervals[index] = (intervals[index][0], max(intervals[index][1], intervals[i][1]))
            else:
                index += 1
                intervals[index] = intervals[i]
        return intervals[:index + 1]

    def part1(self):
        y = 2000000
        result = 0
        intervals = []
        for sensor, beacon in self.sensorstobeacons.items():
            manhattan = abs(sensor[0] - beacon[0]) + abs(sensor[1] - beacon[1])
            y_dist = abs(sensor[1] - y)
            if y_dist > manhattan:
                continue
            x_dist = manhattan - y_dist
            intervals.append((sensor[0] - x_dist, sensor[0] + x_dist))
        intervals = self.merge_intervals(intervals)
        beacon_y = next(beacon for _, beacon in self.sensorstobeacons.items() if beacon[1] == y)
        for interval in intervals:
            result += interval[1] - interval[0] + 1
            if beacon_y[1] == y and interval[0] <= beacon_y[0] <= interval[1]:
                result -= 1
        print(f"Part 1: {result}")

    def part2(self):
        result = 0
        mx = 4000000
        for i in range(mx + 1):
            intervals = []
            for sensor, beacon in self.sensorstobeacons.items():
                manhattan = abs(sensor[0] - beacon[0]) + abs(sensor[1] - beacon[1])
                y_dist = abs(sensor[1] - i)
                if y_dist > manhattan:
                    continue
                x_dist = manhattan - y_dist
                intervals.append((max(0, sensor[0] - x_dist), min(mx, sensor[0] + x_dist)))
            intervals = self.merge_intervals(intervals)
            if len(intervals) > 1:
                result = (intervals[0][1] + 1) * mx + i
                break
        print(f"Part 2: {result}")


if __name__ == "__main__":
    s = Solution()
    file = open("input.txt", "r")
    lines = file.read().splitlines()
    s.init(lines)
    s.part1()
    s.part2()